pkg/io/providers/providers.go

```diff
@@ -0,0 +1,71 @@
+package providers
+
+import (
+	"context"
+	"gocloud.dev/blob"
+	"strings"
+)
+
+const (
+	// Required as long as paths without prefix are used
+	DefaultStorageProviderName = "gs"
+)
+
+var (
+	storageProviderCreators        = map[string]StorageProviderCreator{}
+	storageProviderPathIdentifiers = map[string]StorageProviderPathIdentifiers{}
+)
+
+type StorageProvider interface {
+	ParseStoragePath(storagePath string) (bucket, relativePath string, err error)
+	GetBucket(ctx context.Context, bucketName string) (*blob.Bucket, error)
+	SignedURL(ctx context.Context, bucketName, relativePath string, opts *blob.SignedURLOptions) (string, error)
+}
```

Comment:
Can you explain why this interface has to exist? `ParseStoragePath` is a wrapper around `util.ParseStoragePath` for all three providers, `SignedURL` is a wrapper around `GetBucket().SignedURL()`.

Generally I would love it if we wouldn't have to write code for every provider we want to support. We could for example achieve that by extending prow/flagutil and switch based on auth flag, e.G. none -> directly call `gocloud.dev/blob.OpenBucket`, `--gcs-credentials-file` is set -> read it, constuct a gcp client, call `gcsblob.OpenBucket` etc.

That way we would support all providers that are supported by `gocloud.dev/blob` via their default auth mechanism without having to write any code for it.

Reply:
Hm, I think it comes down to if we can move the code which is different here to other places or if it is so generic that it just works the same for every provider.

* ParseStoragePath
  * it was the same in all 3 providers at the time of your review. But in the meantime (through unit tests) I found out that file must be handled differently. In the s3/gs case we can just strip the prefix. The first segment is the bucket everything else the relative Path. In the file case we have to split in directory and file. Of course this can also be handled by one util and doesn't have to be implemented in the provider

  * GetBucket: works like you described. But this more or less just means putting the code in the flagutil instead of the io package.
  
  * SignedURL: Works the same for s3 and file. Prow currently also uses anonymous access to public GCS buckets, so that differs. Of course the differences can also be handled in a util.

Overall I think we can move each of these 3 methods to 3 functions of the io package (or somewhere else e.g. flagutil). But I think we still have to write code for each provider. I think right now I would prefer keeping the code in the `io` (or `io2`). Because it's similar to the current implementation and it's probably easier to refactor. Of course this doesn't mean we necessarily have to keep the provider interface.

Reply:
Both the special case of `file` not having a bucket portion and `SignedURL` used anonymously can be covered by just using `gocloud.dev/blob.OpenBucket` or not?

I guess my main point is that I do not want all providers to implement some method just because one provider happens to have a special case and that I find the current approach of registering providers way too complicated.

We could for example make the providers export an implementation of
```
type StorageProviderFactory interface {
    AddFlags(fs *flag.FlagSet)
    ValidateFlags() error
    // True if any auth flag for this provider was set
    IsEnabled() bool
    Bucket(path string)(*gocloud.dev/blob.Bucket, error)
}
```

Then iterate over all of them, error if more than one `IsConfigured` and use `gocloud.dev/blob.OpenBucket` if none is configured (e.G. in case of file or unauthenticated gcs).

This allows to keep provider-specifc details like `What auth flags in what combination can be used` and `how do I parse paths` private to themselves and have a fallthrough to the generic `OpenBucket`, allowing to use any supported provider by their default auth method/anonymously.

The other special cases of no prefix => gcs, gcs => gs should IMHO be handled by defaulting (url has no `://` component, prefix it with `gs://`) and replacing (url starts with `gcs://`, replace that with `gs://`)

WDYT?

Reply:
I think I just realized that we have to decide on the following point and then a few things follow from that:

Currently I implemented the package so that we don't have to know which storage provider is used when creating the Opener. So we don't need different flags in the future for GCS/S3/Azure.. and we don't have to add (and propagate) flags for every new provider. We could just use `blob-store-credentials-file` (after a migration from `gcs-credentials-file`). When we then later on access the Opener with a path like `gs://` or `s3://` the correct provider is used based on the path. (see also: https://docs.google.com/document/d/13O1zRX5rpmPCqsPzukavRVqyh4pLp95IeIBVWAZx6Oo/edit?disco=AAAADsYsZxI)

So I think it would be a good idea to decide if we want to have different flags for different providers (or a flag signaling which provider should be used) or just one flag `blob-store-credentials-file`. This would affect the following components:
* deck, gerrit, status-reconciler, tide, gcupload, artifact-uploader, initupload, sidecar
I think this means we have to add the flag(s) to `prowjobs/v1.DecorationConfig` and propagate the flag(s) from the ProwJob to our test Pods. I guess that's why @fejta proposed to [differentiate based on path](https://docs.google.com/document/d/13O1zRX5rpmPCqsPzukavRVqyh4pLp95IeIBVWAZx6Oo/edit?disco=AAAADsYsZxI). I'm not sure if we really should add different flags for different storage providers.


> Both the special case of `file` not having a bucket portion and `SignedURL` used anonymously can be covered by just using `gocloud.dev/blob.OpenBucket` or not?

The difference in parsing the bucket name in `file` could be handled initially by parsing it when using OpenBucket (just fyi `file`
has a bucket portion it's just not the first segment it's the whole directory). I tried OpenBucket via `bucket, err := blob.OpenBucket(context.Background(), "gs://k8s-conform-capi-openstack")`
but it returns the following error: `open bucket gs://k8s-conform-capi-openstack: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.`. I guess that's why this was implemented like this (I only copied the code): https://github.com/kubernetes/test-infra/blob/b46003a27ce81f83c1754bbf1e0cb6765b97287b/pkg/io/v2/providers/gcs/gcs.go#L95-L102


> I guess my main point is that I do not want all providers to implement some method just because one provider happens to have a special case and that I find the current approach of registering providers way too complicated.

> We could for example make the providers export an implementation of
>
> ```
> type StorageProviderFactory interface {
>     AddFlags(fs *flag.FlagSet)
>     ValidateFlags() error
>     // True if any auth flag for this provider was set
>     IsEnabled() bool
>     Bucket(path string)(*gocloud.dev/blob.Bucket, error)
> }
> ```

I refactored it (currently mostly still the same interface) in the export-style and it looks far better imho. Thx! If you want
commit > https://github.com/kubernetes/test-infra/pull/16317/commits/47b2e6b76031013f871b9acb7fda511d99e4aeea
I can also remove the providers now and just implement GetBucket & SignedURL in the `providers.go` once. I implement this in another commit
so I can rollback more easily
commit > upcoming

> Then iterate over all of them, error if more than one `IsConfigured` and use `gocloud.dev/blob.OpenBucket` if none is configured (e.G. in case of file or unauthenticated gcs).

As mentioned above:
 * OpenBucket won't work with unauthenticated GCS
 * Let's decide if we want to use separate flags for separate providers. Then it's a whole other discussion :)

> This allows to keep provider-specifc details like `What auth flags in what combination can be used` and `how do I parse paths` private to themselves and have a fallthrough to the generic `OpenBucket`, allowing to use any supported provider by their default auth method/anonymously.
>
> The other special cases of no prefix => gcs, gcs => gs should IMHO be handled by defaulting (url has no `://` component, prefix it with `gs://`) and replacing (url starts with `gcs://`, replace that with `gs://`)

> WDYT?